File: backend/src/main/java/top/vuhe/auth/JwtTokenFilter.java
package top.vuhe.auth;

import com.baomidou.mybatisplus.core.toolkit.StringUtils;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;
import top.vuhe.common.ApiResponse;
import top.vuhe.common.exception.ExceptionEnum;
import top.vuhe.common.util.ResponseUtils;
import top.vuhe.common.util.TokenUtils;
import top.vuhe.config.SecurityConfig;
import top.vuhe.entity.auth.User;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Arrays;

/**
 * token 认证 Filter
 * <p>
 * 使用token登录(验证用户)
 *
 * @author zhuhe
 */
@Slf4j
public class JwtTokenFilter extends BasicAuthenticationFilter {
    public JwtTokenFilter(AuthenticationManager authenticationManager) {
        super(authenticationManager);
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws IOException, ServletException {
        String url = request.getRequestURI();
        String header = request.getHeader(TokenUtils.AUTHORIZATION);

        //跳过不需要验证的路径
        if (Arrays.asList(SecurityConfig.AUTH_WHITELIST).contains(url)) {
            chain.doFilter(request, response);
            return;
        }

        if (StringUtils.isBlank(header) || !header.startsWith(TokenUtils.TOKEN_PREFIX)) {
            ResponseUtils.send(response, ApiResponse.ofErrorEnum(ExceptionEnum.INVALID_TOKEN));
            return;
        }
        try {
            UsernamePasswordAuthenticationToken authentication = getAuthentication(request);

            SecurityContextHolder.getContext().setAuthentication(authentication);
            chain.doFilter(request, response);
        } catch (Exception e) {
            ResponseUtils.send(response, ApiResponse.ofErrorEnum(ExceptionEnum.INVALID_TOKEN));
            logger.error("Invalid Token " + e.getMessage());
        }
    }

    private UsernamePasswordAuthenticationToken getAuthentication(HttpServletRequest request) {
        String token = request.getHeader(TokenUtils.AUTHORIZATION);
        if (token != null) {
            // 解密Token
            try {
                User user = TokenUtils.validateToken(token);
                return new UsernamePasswordAuthenticationToken(
                        user.getUsername(), null, user.getAuthorities());
            } catch (Exception e) {
                throw new RuntimeException("Invalid Token", e);
            }
        }
        return null;
    }
}

File: backend-kt/src/main/kotlin/top/vuhe/sw/auth/JwtTokenFilter.kt
package top.vuhe.sw.auth

import com.baomidou.mybatisplus.core.toolkit.StringUtils
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter
import top.vuhe.sw.common.ApiResponse
import top.vuhe.sw.common.exception.ExceptionEnum
import top.vuhe.sw.common.util.*
import top.vuhe.sw.config.SecurityConfig
import java.io.IOException
import javax.servlet.FilterChain
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * ## token 认证 Filter
 * 使用token登录(验证用户)
 *
 * @author vuhe
 */
class JwtTokenFilter(authenticationManager: AuthenticationManager) :
    BasicAuthenticationFilter(authenticationManager) {
    @Throws(IOException::class, ServletException::class)
    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, chain: FilterChain) {
        val url = request.requestURI.substring(request.contextPath.length)
        val header = request.getHeader(AUTHORIZATION)

        //跳过不需要验证的路径
        if (SecurityConfig.authWhitelist.contains(url)) {
            chain.doFilter(request, response)
            return
        }
        if (StringUtils.isBlank(header) || !header.startsWith(TOKEN_PREFIX)) {
            send(response, ApiResponse.ofErrorEnum(ExceptionEnum.INVALID_TOKEN))
            return
        }
        try {
            val authentication = getAuthentication(request)
            SecurityContextHolder.getContext().authentication = authentication
            chain.doFilter(request, response)
        } catch (e: Exception) {
            send(response, ApiResponse.ofErrorEnum(ExceptionEnum.INVALID_TOKEN))
            logger.error("Invalid Token " + e.message)
        }
    }

    private fun getAuthentication(request: HttpServletRequest): UsernamePasswordAuthenticationToken? {
        val token = request.getHeader(AUTHORIZATION)
        return if (token != null) {
            // 解密Token
            try {
                val user = validateToken(token)
                UsernamePasswordAuthenticationToken(
                    user.username, null, user.authorities
                )
            } catch (e: Exception) {
                throw RuntimeException("Invalid Token", e)
            }
        } else null
    }
}